mjolnir/core/GBAOABLangevinIntegrator.hpp

```diff
@@ -0,0 +1,338 @@
+
+        std::size_t rattle_step = 0;
+        while(rattle_step<correction_max_iter_)
+        {
+            bool corrected = false;
+            for(std::size_t i=0; i<this->constraints_.size(); ++i)
+            {
+                const auto& constraint = constraints_[i];
+                const auto& indices    = constraint.first;
+                auto& p1  = sys.position(indices[0]);
+                auto& p2  = sys.position(indices[1]);
+                auto& v1  = sys.velocity(indices[0]);
+                auto& v2  = sys.velocity(indices[1]);
+                auto& op1 = this->old_pos_rattle_[indices[0]];
+                auto& op2 = this->old_pos_rattle_[indices[1]];
+                auto& rm1 = sys.rmass(indices[0]);
+                auto& rm2 = sys.rmass(indices[1]);
+
+                const auto      dp = p2 - p1;
+                const real_type dp2 = math::length_sq(dp);
+                const real_type missmatch = square_v0s_[i] - dp2;
+
+                if(abs(missmatch) > tolerance)
+                {
+                    corrected = true;
+                    const auto old_dp = op2 - op1;
+                    const real_type dot_old_new_dp = math::dot_product(old_dp, dp);
+                    const real_type lambda =
+                        0.5 * missmatch * r_reduced_mass_[i] * dot_old_new_dp;
+                    const coordinate_type correction_force = lambda * old_dp;
+                    const coordinate_type correction_vec1  = correction_force * rm1;
+                    const coordinate_type correction_vec2  = correction_force * rm2;
+                    p1 -= correction_vec1;
+                    p2 += correction_vec2;
+                    v1 -= correction_vec1 * r_dt_in_correction_;
+                    v2 += correction_vec2 * r_dt_in_correction_;
+                    op1 = p1;
+                    op2 = p2;
+                }
+            }
+
+            if(!corrected) break;
+
+            ++rattle_step;
+        }
+
+        if(rattle_step >= correction_max_iter_)
+        {
+            MJOLNIR_LOG_WARN("rattle iteration number exceeds rattle max iteration");
+        }
+
+        return;
+    }
+
+    void correct_velocity(system_type& sys, const real_type tolerance)
+    {
+        MJOLNIR_GET_DEFAULT_LOGGER();
+
+        std::size_t rattle_step = 0;
+        while(rattle_step<correction_max_iter_)
+        {
+            bool corrected = false;
+            for(std::size_t i=0; i<this->constraints_.size(); ++i)
+            {
+                const auto& constraint       = constraints_[i];
+                const auto& indices          = constraint.first;
+                auto& p1  = sys.position(indices[0]);
+                auto& p2  = sys.position(indices[1]);
+                auto& v1  = sys.velocity(indices[0]);
+                auto& v2  = sys.velocity(indices[1]);
+                auto& rm1 = sys.rmass(indices[0]);
+                auto& rm2 = sys.rmass(indices[1]);
```

Comment:
reciprocal masses should not be changed, so we need const here.
Also, these seem to be used only in the if block. Move it to the if block.
```suggestion
                const auto rm1 = sys.rmass(indices[0]);
                const auto rm2 = sys.rmass(indices[1]);
```